source/common/config/protobuf_link_hacks.h

```diff
@@ -43,5 +45,8 @@ const envoy::service::listener::v3alpha::LdsDummy _lds_dummy_v3;
 const envoy::service::route::v3alpha::RdsDummy _rds_dummy_v3;
 const envoy::service::cluster::v3alpha::CdsDummy _cds_dummy_v3;
 const envoy::service::endpoint::v3alpha::EdsDummy _eds_dummy_v3;
-const envoy::service::route::v3alpha::SrdsDummy _srds_dummy_v4;
+const envoy::service::route::v3alpha::SrdsDummy _srds_dummy_v3;
+
+// note: hds force linking is not needed because it is still implicitly linked via
```

Comment:
Actually nevermind. According to https://github.com/envoyproxy/envoy/pull/9210 the fact that the v3alpha descriptor is present should be enough to link in the v2 descriptor.

I think #9191's issue with alwayslink in iOS might be at play here, because the iOS build of Envoy Mobile fails at runtime trying to find the descriptors for the metrics service (both the config type and the service method), and more importantly the v2 Bootstrap type.

I fixed locally with this change and by forcing bootstrap v2 linking in this file, and forcing metrics service linking in the extension.

@htuch lmk what you think.

Reply:
cc @keith in case you have a second to take a look, and could provide any insight.

Reply:
Yeah, TBH, I think I'm still confused. I did end up needing to add those dummy messages, despite the claim in https://github.com/protocolbuffers/protobuf/issues/4221 that if we set `always_link`, it shouldn't be needed. This was for a normal Linux build.

However, this only applies to the "proto file with no messages and only service definitions" scenario. What you are describing is different, thing like bootstrap shouldn't be an issue. I think there is a deeper iOS side issue, in addition to the service definition only link issue.

Reply:
What I am thinking is that:

> What you observed is most likely due to the linker dropping unused symbols (i.e., the descriptor registering code is not reachable from your application code so they got removed by linker)

is happening to v2 types for headers that are not explicitly included/used in the Envoy codebase anymore; even though they should be linked because their build targets are included in the v3 bazel targets, and all of them have `always_link`. Hence, we I explicitly use them in the codebase again, even just to declare a dummy variable they are not dropped.

Reply:
Yeah, for bootstrap (for example), the v3alpha proto depends on the v2. It would be good to understand why the linker is doing this to avoid whack-a-mole.

Reply:
I agree that we should understand what the linker is doing in order to avoid whack-a-mole as we include fewer and fewer v2 headers. For the time being, I pushed a potential stopgap addition to the hacks to get us (Envoy Mobile unblocked). We can investigate the linker internally as it is obvious now that the problem is wider than previously thought. wdyt?